---
title: "Outcrossing populations simulations"
author: "[Statistical Genetics Lab](http://statgen.esalq.usp.br) <br/> Department of Genetics <br/> Luiz de Queiroz College of Agriculture <br/> University of SÃ£o Paulo"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Outcrossing Populations}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Simulating the data file 

`Reads2MapTools` offers wrapper functions for [PedigreeSim](https://github.com/PBR/pedigreeSim) (Voorrips and Maliepaard, 2012) software to simulate mapping populations. 

## Run PedigreeSim 

You can run PedigreeSim and use its output files in `Reads2MapTools` or you can use the wrapper function named `run_pedsim` that facilitates this procedure.

The function do not provide every possibility offered by PedigreeSim software. If you want to change any parameter that is not available in the function, please use the PedigreeSim software directly. Here is the [software documentation](https://github.com/PBR/pedigreeSim/tree/master/Manual) for more information.

```{r, eval= FALSE, message=F,warning=F}
# For outcrossing population
run_pedsim(chromosome = "Chr1", n.marker = 54, tot.size.cm = 100, centromere = 50,
           n.ind = 200, mk.types = c("A1", "A2", "A3", "A4", "B1.5", "B2.6", "B3.7", 
                                     "C.8", "D1.9", "D1.10", "D1.11", "D1.12", "D1.13", 
                                     "D2.14", "D2.15", "D2.16", "D2.17", "D2.18"),
           n.types = rep(3,18), pop = "F1", 
           name.mapfile = "mapfile.txt", name.founderfile="founderfile.gen",
           name.chromfile="sim.chrom", name.parfile="sim.par",
           name.out="sim_out")
```

The function allows creating outcrossing, f2 intercross, and backcross populations from bi-parental. You can define it in `pop` argument. You can also define the number of chromosomes (argument `chromosome`), the number of markers in each chromosome (`n.marker`), the total size of the groups in cM (`tot.size.cm`), the position of centromere (`centromere`), number of individuals in the population (`n.ind`), the marker types (`mk.types`, see `OneMap` tutorials) and the number of markers of each type (`n.types`). 

We suggest you open the output files `founderfile`, `chromfile`, `mapfile` and `parfile` to check if they agree with your intentions before proceed to other analysis.

## Convert PedigreeSim output to `OneMap` raw data

Once you run PedigreeSim (directly or using our `run_pedsim` function), you should have the output file `genotypes.dat` (see an example file called `sim_cod_out_genotypes.dat` in `OneMap` extdata directory). To convert this to `OneMap` raw file, you just need to specify the cross type (`cross`), which ones are the parents (`parent1` and `parent2`) and if you want to include missing genotypes (`miss.perc`). Only cross types `outcross` and `f2 intercross` are supported by now.

```{r, eval=FALSE, message=F,warning=F}
# For outcrossing population
pedsim2raw(genofile = "sim_out_genotypes.dat", cross="outcross", parent1 = "P1", parent2 = "P2", out.file = "sim_out.example.raw", miss.perc = 0)
```

## Convert PedigreeSim output to VCF file 

The same output file from PedigreeSim, the `genotypes.dat` can be used to simulate a VCF file together with the PedigreeSim `mapfile` and `chrom`. The advantages to simulate a VCF instead of `OneMap` raw file is that VCF is a standard file format and can store a lot of other information included the allele counts (usually in the field `AD` or `DPR`). The `pedsim2vcf` function can simulate the allele counts using negative binomial or updog distributions (argument `method`). The main parameters for the distributions are defined with arguments `mean.depth`, that defines the mean allele depth in the progeny, `p.mean.depth` that defines the mean allele depth in the parents, argument `disper.par` defines the dispersion parameter, `mean.phred` sets the mean Phred score of the sequencing technology used. The function can also simulate missing data (`miss.perc`). Through arguments `pos` and `chr` you can define vectors with physical position and chromosome of each marker. With argument `haplo.ref` you specify which one of the haplotypes in `genotypes.dat` will be considered the reference. Establishing `phase` as TRUE, the VCF will have phased genotypes. After allele counts are simulated, the genotypes are re-estimated using a binomial distribution. The VCF generated by this function only has one or two FORMAT fields, the GT and AD (if `counts = TRUE`). **Dominant markers are not supported by this function**.

```{r, eval=FALSE, message=F,warning=F}
# Dominant markers are not supported, then, we simulate other dataset with only codominant markers
run_pedsim(chromosome = "Chr1", n.marker = 42, tot.size.cm = 100, centromere = 50,
           n.ind = 200, mk.types = c("A1", "A2", "B3.7", "D1.9", "D1.10", "D2.14", "D2.15"),
           n.types = rep(6,7), pop = "F1", 
           name.mapfile = "mapfile_out.txt", name.founderfile="founderfile.gen",
           name.chromfile="sim_out.chrom", name.parfile="sim.par",
           name.out="sim_cod_out")
```

```{r, eval=FALSE, message=F,warning=F}
# For outcrossing population
pedsim2vcf(inputfile = "sim_cod_out_genotypes.dat", 
           map.file = "mapfile_out.txt", 
           chrom.file = "sim_out.chrom", 
           out.file = "simu_cod_out.vcf",
           miss.perc = 0, 
           counts = TRUE, 
           mean.depth = 100, 
           p.mean.depth = 100, 
           chr.mb = 10, 
           method = "updog", 
           mean.phred = 20,
           bias=1, 
           od=0.00001,
           pos=NULL,
           chr=NULL,
           phase = FALSE,
           disper.par=2)
```

The function prints on screen the percentage of genotypes changed between simulated by PedigreeSim and re-estimated using a binomial distribution and simulated allele counts. Notice that this percentage increase if you reduce the mean depth or increase bias or dispersion parameter.

## Comparing estimated and simulated haplotypes

The simulations here included are very useful to test the efficiency of our algorithms, we made them available in case to be useful for other purposes. Using functions `run_pedsim`, `pedsim2vcf` with argument phased `TRUE`, we can convert the phased VCF using `vcf2progeny_haplotypes`, which keeps the original haplotype. With that, we can compare the haplotypes simulated and the ones estimated by the onemap approach.

Here we will use the file `sim_out_cod_genotypes.dat` to get one phased VCF. See that now we don't want to simulate read counts or genotype errors; we just want the original haplotypes in VCF format, then we set `counts = FALSE` and don't need to define the related arguments.

```{r, eval=FALSE}
pedsim2vcf(inputfile = system.file("extdata/sim_cod_out_genotypes.dat", package = "onemap"),
           map.file = system.file("extdata/mapfile_out.txt", package = "onemap"), 
           chrom.file = system.file("extdata/sim_out.chrom", package = "onemap"), 
           out.file = "simu_out_phased.vcf",
           miss.perc = 0, 
           counts = FALSE, 
           chr.mb = 10,
           pos="cM",
           chr=NULL,
           phase = TRUE)
```

Now, we can convert the phased VCF to onemap_progeny_haplotypes object using function `vcf2progeny_haplotypes`. The position, the group names, and individuals id will be according to the ones contained in VCF file. This function can take some time to run if you select too many individuals or groups.

```{r, eval=FALSE}
vcfR.object <- read.vcfR("simu_out_phased.vcf")

progeny_dat <- vcf2progeny_haplotypes(vcfR.object = vcfR.object, ind.id = c("F1_004", "F1_005"), group_names = "Chr1", parent1 = "P1", parent2 = "P2", crosstype = "outcross")
plot(progeny_dat)
```

This is the simulated dataset. If OneMap algorithm could get all information needed, it could reproduce these haplotypes exactly how they were in simulated data. Let's see how OneMap estimates these haplotypes:

```{r, eval=FALSE}
onemap_test <- onemap_read_vcfR(vcfR.object = vcfR.object, cross = "outcross", parent1 = "P1", parent2 = "P2", only_biallelic = FALSE)

onemap_test # Just to remember of which dataset we are talking about

twopts <- rf_2pts(onemap_test)
seq1 <- make_seq(twopts, "all")
map_test <- map(seq1)
progeny_est <- progeny_haplotypes(map_test, ind = c(4,5), most_likely = FALSE)
plot(progeny_est)
```

See that, besides markers are not in their exact position, the recombination breakpoints are the same as the ones simulated. The marker position could only be the same as the ones simulated if we have a very big (or maybe infinite) population size. Still, here we have only 200 individuals, but this is not a problem once we could reproduce the haplotypes.